File: 207. 课程表.py
import collections


class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        in_degree = [0] * numCourses  # 记录课程需要几节前置课
        adj = collections.defaultdict(list)  # 记录学完这门课程能解锁什么课
        for target, pre in prerequisites:
            adj[pre].append(target)
            in_degree[target] += 1

        # 无前置课程的课入队
        q = collections.deque()
        for i in range(numCourses):
            if in_degree[i] == 0:
                q.append(i)
        learned_count = 0

        # 开始上课，出队的课就是上完的
        while q:
            curr_course = q.popleft()
            learned_count += 1
            for next_course in adj[curr_course]:
                in_degree[next_course] -= 1  # 更新后续课程所需的前置课数量
                if in_degree[next_course] == 0:  # 如果前置课上完了，就入队
                    q.append(next_course)
        return learned_count == numCourses

File: 207. 课程表.go
package main

func canFinish(numCourses int, prerequisites [][]int) bool {
	graph := make([][]int, numCourses)  // 上完这个课能解锁什么课
	inDegree := make([]int, numCourses) // 这个课需要几门前置课

	for _, p := range prerequisites {
		course, pre := p[0], p[1]
		graph[pre] = append(graph[pre], course)
		inDegree[course]++
	}

	q := make([]int, 0)
	for i := range numCourses {
		if inDegree[i] == 0 {
			q = append(q, i)
		}
	}

	count := 0
	for len(q) > 0 {
		curr := q[0]
		q = q[1:] // 上完课就出队
		count++

		for _, nextCourse := range graph[curr] {
			inDegree[nextCourse]--         // 他解锁的课，前置课的数量减少
			if inDegree[nextCourse] == 0 { // 如果减少到0了，就入队准备上课
				q = append(q, nextCourse)
			}
		}
	}

	return count == numCourses
}
